### PR TITLE
External schemas auto discovery

### PR DESCRIPTION
**Pull request by Thorium**
I started to look into external definition loading in Parsers.fs, but I need a bit help here.

1) The links to external refs do have relative paths. These paths could be http or local. The main swagger-provider has resolved the basePath of the root schema file, but there is no property near to expose it. What if the file is not found, how complex resolving of the paths should be implemented?

2) When the file content is gotten, it should be parsed recursively. But the parsers needed are defined later in the process. There are many possible ways to continue from here:
   - Make the logic recursive. Sadly, it's not just `parsePathsObject` and `parseSwaggerObject`, because the adapter files are also needed, and they are located in different module. Would need a re-ordering and splitting the logic.
   - Or extend `OperationObject` to have a to-do-later special property that would be parsed later. This would avoid recursions, but it seems that the object is used widely, so it's more work.
   - Or copy&paste the minimal parsing logic, code-duplication and hacky-feeling. Wouldn't solve more complex scenarios.
   - Or create totally another instance of another SwaggerProvider for handling the linked schema file. I'm expecting this would lead to some-level high memory-usage, and possibly reflection, because the schemas could have references to each other.

### CODE REVIEW

**Comment by Thorium** [comment]
Currently this PR has a little code clean-up and it adds an ignored non-supported test, so it could be merget, but it doesn't solve the real issue.

**Comment by Thorium** [comment]
External schemas auto discovery is now done for the SwaggerProvider. 
Should be working but not tested properly as my schema was OpenApi, not Swagger.

The problem 2 mentioned above: the parser is transferred as a parameter to parseSwaggerObject. parsePathsObject will return a list of Swagger-objects that can be added to the type provider types. 

I also added a static parameter ResolveReferences that will have to be set true to use this external schema discovery. The reasons for this are: 1) maybe people don't want the discovery due to security reasons, what if the schema is referring a malicious source like a virus? 2) Also if the schema is referring to 100MB file that would jam your computer, and you want 100% backward-compatibility not to explore the external schema.

Sadly, I didn't resolve my use-case of using external yaml schemas with OpenApi provider, and the reason seems to be Microsoft.OpenApi.Readers not supporting refs-reading just yet:
https://github.com/microsoft/OpenAPI.NET/issues/288 I hope this will be automatically fixed with this PR after they release a stable 1.2 version and we move to use that.

**Comment by sergey-tihon** [comment]
TBH, I did not plan to improve SwaggerProvider further. I hope that now day we  just delete it and `OpenApiClientProvider` cover all cases. I decided to keep for a while, because during active work on OpenApi support I find out that there are many 'old' v2 schemas that `Microsoft.OpenApi.Readers` cannot parse.

How hard would be to pre-process parsed schema, find them, load, merge into one schema, repeat until we build schema without external references.

>I also added a static parameter ResolveReferences that will have to be set true to use this external schema discovery. 

I like this approach in general. If `ResolveReferences` is `false` TP fails when find external ref and tell user that he should set it `true` if he trust to external links.

**Comment by sergey-tihon** [comment]
Have you tried any tool that can merge multiple spec files into one? Maybe it will be easier to create one mega schema file and then use TP (and faster in runtime)

Quick search showed that there is a tool called [speccy](https://github.com/wework/speccy) that can merge multiple OpenAPI files into one
> resolve [options] <file-or-url>  pull in external $ref files to create one mega-file

here is related blog post https://blog.runscope.com/posts/how-to-merge-openapi-definition-files

**Comment by Thorium** [comment]
Yes I tried, it failed, the Microsoft.OpenApi reader still gives errors (even though paritally result is ok), maybe because the schema files contains references to each other, and speccy creates weird links like:
```yaml
      responses:
        "200":
          description: Generated
          content:
            application/json:
              schema:
                $ref: "#/paths/~1sdk_token/post/requestBody/content/application~1json/schema"
        default:
          $ref: "#/paths/~1applicants/post/responses/default"
```

I'm trying to read https://github.com/onfido/onfido-openapi-spec

**Comment by Thorium** [comment]
No offence to M$, but I kind of doubt if their C# schema-parsing-project will ever be better than your few lines of F#.

File: src/SwaggerProvider.DesignTime/Provider.SwaggerClient.fs [hunk 1]
```diff
@@ -61,27 +63,40 @@ type public SwaggerTypeProvider(cfg : TypeProviderConfig) as this =
...
+                        defCompiler.Namespace.GetProvidedTypes()
+
+                    let tys = compile mainSchema
+
+                    let rec fetchExternalSchemaTypes (schema:Schema.SwaggerObject) (collected:ProvidedTypeDefinition list) =
+                        if Array.isEmpty schema.SwaggerObjects then collected
+                        else schema.SwaggerObjects |> Array.toList |> List.map(fun eSchema ->
+                                  compile eSchema |> fetchExternalSchemaTypes eSchema) |> List.concat
+
+                    let externalTys = fetchExternalSchemaTypes mainSchema []
```
Comment (sergey-tihon) [inline src/SwaggerProvider.DesignTime/Provider.SwaggerClient.fs:92]: Can  we move this line inside `if resolveReferences then` and do not fetch external types when user does not request it.

File: src/SwaggerProvider.DesignTime/v2/Parser/Parsers.fs [hunk 2]
```diff
@@ -401,18 +401,69 @@ module Parsers =
...
+                    let fname = obj.AsString()
+                    (if fname.Contains("#") then fname.Split('#').[0] else fname)
+                let execAssembly =
+#if INTERACTIVE
+                    "" //not supported
+#else
+                    System.Reflection.Assembly.GetExecutingAssembly().Location
+#endif
+                let tryPaths =
+                    [ fileName;
```
Comment (sergey-tihon) [inline src/SwaggerProvider.DesignTime/v2/Parser/Parsers.fs:446]: Why do not you use `Schema` TP parameter (url or file path) without filename as an option (1st option) as base.
For example if we  build TP for `http://my.server.com/v2/swagger.yaml` and found `../v2/token.yaml#/token` it most likely refer to type from `http://my.server.com/v2/token.yaml`

**Pull request closed without merging**